File: nmLabReport/lab16.c
#include <stdio.h>
#include <math.h>

#define MAX_ITER 100
#define TOL 1e-6

void jacobi(int n, double a[n][n], double b[n], double x[n]) {
    double x_new[n];
    int iter = 0;
    double error;

    do {
        error = 0.0;
        for (int i = 0; i < n; i++) {
            x_new[i] = b[i];
            for (int j = 0; j < n; j++) {
                if (i != j) {
                    x_new[i] -= a[i][j] * x[j];
                }
            }
            x_new[i] /= a[i][i];
            error += fabs(x_new[i] - x[i]);
        }

        for (int i = 0; i < n; i++) {
            x[i] = x_new[i];
        }

        iter++;
    } while (error > TOL && iter < MAX_ITER);
}

int main() {
    int n;
    printf("Enter the number of equations: ");
    scanf("%d", &n);

    double a[n][n], b[n], x[n];

    printf("Enter the coefficients of the matrix A:\n");
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            scanf("%lf", &a[i][j]);
        }
    }

    printf("Enter the constants vector B:\n");
    for (int i = 0; i < n; i++) {
        scanf("%lf", &b[i]);
        x[i] = 0; // Initial guess
    }

    jacobi(n, a, b, x);

    printf("The solution is:\n");
    for (int i = 0; i < n; i++) {
        printf("x[%d] = %lf\n", i, x[i]);
    }

    return 0;
}

File: nmLabReport/lab16.py
import numpy as np

def jacobi(a, b, x, tol=1e-6, max_iter=100):
    n = len(a)
    x_new = np.zeros_like(x)
    for _ in range(max_iter):
        for i in range(n):
            s = sum(a[i][j] * x[j] for j in range(n) if j != i)
            x_new[i] = (b[i] - s) / a[i][i]
        if np.allclose(x, x_new, atol=tol):
            break
        x = np.copy(x_new)
    return x

def main():
    n = int(input("Enter the number of equations: "))
    a = np.zeros((n, n))
    b = np.zeros(n)
    x = np.zeros(n)

    print("Enter the coefficients of the matrix A:")
    for i in range(n):
        a[i] = list(map(float, input().split()))

    print("Enter the constants vector B:")
    b = list(map(float, input().split()))

    x = jacobi(a, b, x)

    print("The solution is:")
    for i in range(n):
        print(f"x[{i}] = {x[i]:.6f}")

if __name__ == "__main__":
    main()